### DISCUSSION TITLE
Categories/Folders

### DISCUSSION THREAD

**Discussion by mgerasolo** [daya0576/beaverhabits#52] [Ideas]
It would be great if there were categories/folders.  This could either be done by making actual toggle/collapsible header rows with a drag and drop interface.  And then the folder/category would have a circle chart for how many of the items are completed.

Or there it is probably easier from a code level to just let people have multiple lists on the same page.  For example I can have a Diet List and Exercise List, and Self-Improvement List on the same page.  I would imagine that would be a simpler level of effort.

**Comment by daya0576** [comment]
That's a great idea. Is the *category* mainly for static analysis or for completing habits on a single page?

I am planning to add a page that displays statistics for a single category or multiple habits. However, I'm not sure if this is exactly what you want.

**Comment by mgerasolo** [comment]
Categories would have 2 functions:
  1) Cluster related items either by topic (food, exercise, etc) or by timing (Morning, Afternoon, Night)
  2) Ideally a role up of the related items (I like the idea of using a circle chart to represent how much of it is complete).
  
  I was using something similar for a few months using poster board and push pins for 15 days at a time.  And then moved to a Google Sheet type technique.  But I think your app is a cleaner alternative with better data aggregation and trend tracking.  The attached photos help show how I was using Headers to cluster items.
  
  I also had multiple states unknown (grayed out), passed (green), failed (red), Exception (gray box).  The Exception is for things that I did not yet have started at that point, or if I did an excessive amount of work around the house that was physical, then it was an exception on the exercise type activities.  Or if I was sick or injured then exception to mark it as not complete but not a fail.  Some of my goals may seem basic, but I am partly handicapped.
  
  I also had targets (like 80% of the days for Clean Eating) as many habits are not necessarily every single day, as you need days to rest or enjoy indulging.
  
  By circle charts, I mean something like the attached view from StepsApp which shows how many steps per day vs the target step count)(
  
  
![IMG_4896](https://github.com/user-attachments/assets/618755ae-3132-458a-8eb4-c8220a4313cc)
![Screenshot 2024-12-31 at 12 26 47 AM](https://github.com/user-attachments/assets/c77ca68d-2448-4186-9e10-ab98cc2cdc6f)
![IMG_4893](https://github.com/user-attachments/assets/1ee4f77e-e07f-40d0-83f6-51bca209b59e)

**Comment by daya0576** [comment]
I have broken down the requirements into the following parts:
1. Adding a new habit type with multiple states
2. Adding tags to habits, such as topic and timing
3. Adding a “dashboard” to display overall completion rates

Actually, it needs a lot of effort to implement these "features"

But thank you for the detailed input. When I have time, I will work on completing this ultimate version of the web app :p

**Reply by ThePhoenixRiseth** [reply]
I fully get all of these being major work. Just wanted to at least show some ideas for similar systems I liked and thought helpful. As it might inspire the direction for where you may want to go as a whole.

At the least the grouping concept whether it be folder or having multiple lists on the same page seemed like a reasonable size improvement to start with.

**Reply by daya0576** [reply]
Sure, I have created several issues to track the improvements: 
- #53 
- #54

I will inform you as soon as the feature is available :)

**Comment by nov1n** [comment]
While I understand the motivation behind this feature request, I have some reservations about potentially deviating too far from the project's original vision and design philosophy. One of the main selling points that drew me to this application over more feature-rich alternatives is its simple, uncomplicated approach.

Personally, I would be somewhat put off if the app strays too much from its original vision by getting overloaded with an abundance of new features and "goals."

Of course, this is just my own perspective, but I wanted to respectfully voice my stance in hopes of counterbalancing potential feature creep with a reminder of the attractive core principles that drew many of us to adopt this tool initially.

**Reply by mgerasolo** [reply]
I fully understand the concern with feature creep and its simplicity is its strength in a lot of ways.  Thats why I originally only submitted the adding folders suggestion (or ability to have multiple lists on the same page if you want to keep styling like it is now if that is easier).  And it would be up to the user whether they make a folder/second list or not.  So there is no loss of its simplicity with that option.

Adding later the option to summarize the list/folder with a circle graph would also I think be in line with the original simplicity if the level of effort is not too much.

The other features I think would help, but definitely should be considered only if they can fit in with the original vision of keeping it simple.  It is up to the project leads to decide if they want to include, but I figured I would bring it up as one of the pain points on a system like this is it is easy to not be sure if you explicitly did not do a habit or if you just forgot to mark it as complete.  When looking at trends and such if you want any of the numbers to be meaningful making the distinction between marking it as not done or forgetting to mark it after doing it I find to be very helpful.  For example this is already an app for people who are not good with building habits and looking for tools to do so.  So unless you have the habit of making sure to mark everything at the end of the night (which is one of the habits on my tracker), then you sit there and go did I not do the habit, or did I just forget to mark that I did it.

**Reply by daya0576** [reply]
Thanks for all the input, I totally agree that the app should be as minimal as possible. 

And I do appreciate each user's feature requests and pain points, which have made beaverhabits great over the past few months.

Here are strategies to avoid feature creep:
1. Global feature flags: the default setup should be simple and clean, but feature flags can be enabled depending on user preferences, for example, the count badge for the index page. Actually, I was struggling to implement this user request feature in the beginning, and spent days refining the style to make it natural. Eventually, I'm satisfied with this feature, making the app unique and different from "loop habit". 
2. Extended habit type: Currently there is only one habit type, maybe in the future, new habit types can be extended to provide additional behaviours. Not sure how to achieve it now, but it is definitely optional for the user :p
